Use `Scratch` for cached timezone data.

MichaelHatherly opened the pull request:
Fixes #338, based on https://github.com/JuliaTime/TimeZones.jl/issues/343#issuecomment-848823981 by KristofferC.

At module init runs the current `build` to generate
the cached data, and stores it in a scratchspace instead of the
deps folder. Associated globals that were defined at top-level are
now set in `__init__` (and `_init` for submodules) to allow for
`PackageCompiler` support. Scratchspace is linked to the Julia and
package version so should rebuild when those change for users.

The `deps/build.jl` now just imports the package which triggers
a build of the cache if needed so this shouldn't really impact
`using TimeZones` much as far as I could tell.

A required PR for `Scratch` to expand it's version support is at
https://github.com/JuliaPackaging/Scratch.jl/pull/24. I've enabled
support going back to 1.0 for that package to match what `TimeZones`
says it supports, but it turns out `EzXML` is bounded at `1.3` so
I couldn't test down to 1.0-1.2.

omus commented on src/TimeZones.jl:
```diff
@@ -30,14 +32,36 @@ export TimeZone, @tz_str, istimezone, FixedTimeZone, VariableTimeZone, ZonedDate
 function __init__()
+    global DEPS_DIR = @get_scratch!("timezones-$VERSION-$pkg_version")
+    TZData._init()
+    Sys.iswindows() && WindowsTimeZoneIDs._init()
```
```suggestion
    @static Sys.iswindows() && WindowsTimeZoneIDs._init()
```

MichaelHatherly replied:
Const-prop should handle that one fine I'd assume and remove the branching by itself. I can add a `@static` anyway if it's preferred though.

omus commented on src/TimeZones.jl:
```diff
@@ -30,14 +32,36 @@ export TimeZone, @tz_str, istimezone, FixedTimeZone, VariableTimeZone, ZonedDate
+    global DEPS_DIR = @get_scratch!("timezones-$VERSION-$pkg_version")
+    TZData._init()
+    Sys.iswindows() && WindowsTimeZoneIDs._init()
+    if isempty(readdir(DEPS_DIR))
+        cd(DEPS_DIR) do
+            for (path, contents) in DEPS_CONTENTS
+                mkpath(dirname(path))
+                write(path, contents)
+            end
+        end
+        build()
+    end
```
I suspect it would be cleaner to just have the build process write the scratch directory?

MichaelHatherly replied:
Yes, I'll move it into `build()` instead.

omus commented on src/TimeZones.jl:
```diff
@@ -30,14 +32,36 @@ export TimeZone, @tz_str, istimezone, FixedTimeZone, VariableTimeZone, ZonedDate
 # abstract type UTC <: TimeZone end  # Already defined in the Dates stdlib
 abstract type Local <: TimeZone end
 
+const pkg_version = VersionNumber(TOML.parsefile(joinpath(dirname(@__DIR__), "Project.toml"))["version"])
```
```suggestion
const PKG_VERSION = VersionNumber(TOML.parsefile(joinpath(dirname(@__DIR__), "Project.toml"))["version"])
```

omus reviewed (commented):
Thanks for this!

omus commented on src/TimeZones.jl:
```diff
@@ -30,14 +32,36 @@ export TimeZone, @tz_str, istimezone, FixedTimeZone, VariableTimeZone, ZonedDate
+const pkg_version = VersionNumber(TOML.parsefile(joinpath(dirname(@__DIR__), "Project.toml"))["version"])
+
 function __init__()
+    global DEPS_DIR = @get_scratch!("timezones-$VERSION-$pkg_version")
```
May be able to use the tzdata version

MichaelHatherly replied:
Can do, is the cache folder structure likely to remain the same between package versions? If not then using just `tzdata_version` as the scratch key may cause issues, perhaps.

omus replied:
It's probably safest to include Julia version, package version, and tzdata version. It also just occurred to me that users are allowed to [re-compile the time zones to support dates in the far future](https://juliatime.github.io/TimeZones.jl/stable/faq/#future_tzs-1). I'll need to think more about this.

omus commented:
This change will require us to drop support for Julia versions below 1.5. That's not a deal breaker but annoying as I want to continue to support the 1.0 LTS while it's the official LTS. The changes for dropping 1.0 support can be done in another PR. We'll need to at least disable the 1.0 CI jobs though so we can validate the changes.

MichaelHatherly commented:
I'd like to retain LTS support as well, hence https://github.com/JuliaPackaging/Scratch.jl/pull/24 should be completed before finishing off this one, so that we can just use a version of `Scratch` that supports Julia `1+`.

MichaelHatherly commented:
@omus the CI failure is a bit of a mystery to me.. I've adjusted some paths and checks so maybe it'll get by this time.

omus commented on src/tzdata/TZData.jl:
```diff
@@ -14,11 +14,17 @@ end
-const TZ_SOURCE_DIR = joinpath(DEPS_DIR, "tzsource")
-const COMPILED_DIR = joinpath(DEPS_DIR, "compiled", string(VERSION))
+function _init()
+    global ARCHIVE_DIR = joinpath(TimeZones.DEPS_DIR, "tzarchive")
+    global TZ_SOURCE_DIR = joinpath(TimeZones.DEPS_DIR, "tzsource")
+    global COMPILED_DIR = joinpath(TimeZones.DEPS_DIR, "compiled", string(VERSION))
+    global ACTIVE_VERSION_FILE = joinpath(TimeZones.DEPS_DIR, "active_version")
+    global LATEST_FILE = joinpath(TimeZones.DEPS_DIR, "latest")
```
I think I'd prefer if these were all global constants and we used `Ref{String}()`

MichaelHatherly replied:
Are any of these constants regarded as public? It'll of course be a breaking change if that's the case unfortunately.

omus commented on src/tzdata/build.jl:
```diff
@@ -43,11 +43,7 @@ function build(
-                if tzdata_hash === nothing
-                    error("Latest tzdata is $latest_version which is not present in the Artifacts.toml")
-                end
+                @artifact_str "tzdata$latest_version"
```
Did this need to be changed?

MichaelHatherly replied:
It should have the same effect of throwing an error when the particular artifact does not exist, but by using `@artifact_str` the `Artifact.toml` contents gets embedded rather than looked up at runtime in the file, which may not exist if it's been relocated.

omus commented on deps/build.jl:
```diff
@@ -1,3 +1 @@
-using TimeZones: build
-
-build()
+using TimeZones
```
Hmmm, if someone manually runs `] build` we should probably re-build the scratch directory.

andreasnoack commented:
Are there any unaddressed comments here or is this good to go?

MichaelHatherly commented:
> Are there any unaddressed comments here or is this good to go?

The failure in benchmarks is the current blocker here, still trying to pin it down.

MichaelHatherly commented:
Closing in favour of #348.
